### PR TITLE
[Merged by Bors] - feat: port Analysis.Convex.Function

### PR DESCRIPTION
**Pull request by mcdoll**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Convex/Function.lean [hunk 1]
```diff
@@ -0,0 +1,1249 @@
...
+theorem LinearOrder.convexOn_of_lt (hs : Convex 𝕜 s)
+    (hf : ∀ ⦃x⦄, x ∈ s → ∀ ⦃y⦄, y ∈ s → x < y → ∀ ⦃a b : 𝕜⦄, 0 < a → 0 < b → a + b = 1 →
+      f (a • x + b • y) ≤ a • f x + b • f y) :
+    ConvexOn 𝕜 s f := by
+  refine' convexOn_iff_pairwise_pos.2 ⟨hs, fun x hx y hy hxy a b ha hb hab => _⟩
+  wlog h : x < y
+  · rw [add_comm (a • x), add_comm (a • f x)]
+    rw [add_comm] at hab
+    -- `wlog` gives a bad term, which has picked up all the stray variables.
+    -- See https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/wlog.20.2316495
```
Comment (eric-wieser) [inline Mathlib/Analysis/Convex/Function.lean:456]: I think using `wlog generalizing` would be better than listing every variable here!

**Comment by mcdoll** [comment]
I don't know what to do about the excessive use of `etaExperiment`. Is it better to get rid of them (i.e., try to find another hack to work around 2074) or keep it there and deal with it once lean4 has a solution?

**Comment by Parcly-Taxel** [comment]
> I don't know what to do about the excessive use of `etaExperiment`.

It is better to keep the eta enablers there.

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**